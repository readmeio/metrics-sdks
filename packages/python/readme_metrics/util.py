import logging
from logging import Formatter, StreamHandler


def util_build_logger():
    logger = logging.getLogger(__package__)
    logger.setLevel(logging.CRITICAL)
    formatter = Formatter(fmt="%(asctime)s [%(levelname)s] %(message)s")
    handler = StreamHandler()
    handler.setFormatter(formatter)
    logger.addHandler(handler)
    return logger
