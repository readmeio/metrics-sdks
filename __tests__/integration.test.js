import http from 'http';
import { cwd } from 'process';
import { spawn } from 'child_process';
import { promisify } from 'util';
import { once } from 'events';
import getPort from 'get-port';
import caseless from 'caseless';

if (!process.env.EXAMPLE_SERVER) {
  // eslint-disable-next-line no-console
  console.error('Missing `EXAMPLE_SERVER` environment variable');
  process.exit(1);
}

// https://gist.github.com/krnlde/797e5e0a6f12cc9bd563123756fc101f
http.get[promisify.custom] = function getAsync(options) {
  return new Promise((resolve, reject) => {
    http
      .get(options, response => {
        response.end = new Promise(res => {
          response.on('end', res);
        });
        resolve(response);
      })
      .on('error', reject);
  });
};

const get = promisify(http.get);

const randomApiKey = 'a-random-readme-api-key';

// Converts an array of headers like this:
// [
//   { name: 'host', value: 'localhost:49914' },
//   { name: 'connection', value: 'close' },
// ];
//
// To an object that can be passed in to caseless:
// {
//    host: 'localhost:49914',
//    connection: 'close'
// }
function arrayToObject(array) {
  return array.reduce((prev, next) => {
    return Object.assign(prev, { [next.name]: next.value });
  }, {});
}

describe('Metrics SDK Integration Tests', () => {
  let metricsServer;
  let httpServer;
  let PORT;

  beforeAll(async () => {
    metricsServer = http.createServer().listen(0, 'localhost');

    await once(metricsServer, 'listening');
    const { address, port } = metricsServer.address();
    PORT = await getPort();

    // In order to use child_process.spawn, we have to provide a
    // command along with an array of arguments. So this is a very
    // rudimental way of splitting the two values provided to us
    // from the environment variable.
    //
    // I tried refactoring this to use child_process.exec, which just
    // takes in a single string to run, but that creates it's own
    // shell so we can't do `cp.kill()` on it later on (because that
    // just kills the shell, not the actual command we're running).
    //
    // Annoyingly this works under macOS, so it must be a platform
    // difference when running under docker/linux.
    const [command, ...args] = process.env.EXAMPLE_SERVER.split(' ');

    httpServer = spawn(command, args, {
      cwd: cwd(),
      env: {
        PORT,
        METRICS_SERVER: new URL(`http://${address}:${port}`).toString(),
        README_API_KEY: randomApiKey,
        ...process.env,
      },
    });
    return new Promise((resolve, reject) => {
      httpServer.stderr.on('data', data => {
        // eslint-disable-next-line no-console
        console.error(`stderr: ${data}`);
        return reject(data.toString());
      });
      httpServer.on('error', err => {
        // eslint-disable-next-line no-console
        console.error('error', err);
        return reject(err.toString());
      });
      // eslint-disable-next-line consistent-return
      httpServer.stdout.on('data', data => {
        if (data.toString().match(/listening/)) return resolve();
        // eslint-disable-next-line no-console
        console.log(`stdout: ${data}`);
      });
    });
  });

  afterAll(() => {
    httpServer.kill();
    return new Promise((resolve, reject) => {
      metricsServer.close(err => {
        if (err) return reject(err);
        return resolve();
      });
    });
  });

  // TODO this needs fleshing out more with more assertions and complex
  // test cases, along with more servers in different languages too!
  it('should make a request to a metrics backend with a har file', async () => {
    await get(`http://localhost:${PORT}`);

    const [req] = await once(metricsServer, 'request');
    expect(req.url).toBe('/v1/request');
    expect(req.headers.authorization).toBe('Basic YS1yYW5kb20tcmVhZG1lLWFwaS1rZXk6');

    let body = '';
    // eslint-disable-next-line no-restricted-syntax
    for await (const chunk of req) {
      body += chunk;
    }
    body = JSON.parse(body);
    const [har] = body;

    // Check for a uuid
    // https://uibakery.io/regex-library/uuid
    // eslint-disable-next-line no-underscore-dangle
    expect(har._id).toMatch(/^[0-9a-f]{8}-[0-9a-f]{4}-[0-5][0-9a-f]{3}-[089ab][0-9a-f]{3}-[0-9a-f]{12}$/i);
    expect(har.group).toMatchSnapshot();
    expect(har.clientIPAddress).toBe('127.0.0.1');

    const { request, response } = har.request.log.entries[0];

    expect(request.url).toBe(`http://localhost:${PORT}/`);
    expect(request.method).toBe('GET');
    expect(request.httpVersion).toBe('HTTP/1.1');

    const requestHeaders = caseless(arrayToObject(request.headers));
    expect(requestHeaders.get('connection')).toBe('close');
    expect(requestHeaders.get('host')).toBe(`localhost:${PORT}`);

    expect(response.status).toBe(200);
    expect(response.statusText).toBe('OK');
    expect(response.content.text).toBe(JSON.stringify({ message: 'hello world' }));
    expect(response.content.size).toBe(25);
    expect(response.content.mimeType).toBe('application/json; charset=utf-8');

    const responseHeaders = caseless(arrayToObject(response.headers));
    expect(responseHeaders.get('content-type')).toBe('application/json; charset=utf-8');
  });
});
