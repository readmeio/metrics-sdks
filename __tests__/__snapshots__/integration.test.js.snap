// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metrics SDK Integration Tests should make a request to a metrics backend with a har file 1`] = `
Object {
  "email": "owlbert@example.com",
  "id": "owlbert-api-key",
  "label": "Owlbert",
}
`;

exports[`Metrics SDK Integration Tests should make a request to a metrics backend with a har file 2`] = `
Object {
  "cache": Object {},
  "pageref": "http://localhost/",
  "request": Object {
    "bodySize": 0,
    "cookies": Array [],
    "headers": Array [
      Object {
        "name": "host",
        "value": "localhost",
      },
      Object {
        "name": "connection",
        "value": "close",
      },
    ],
    "headersSize": 0,
    "httpVersion": "HTTP/1.1",
    "method": "GET",
    "queryString": Array [],
    "url": "http://localhost",
  },
  "response": Object {
    "bodySize": 0,
    "content": Object {
      "mimeType": "application/json; charset=utf-8",
      "size": 25,
      "text": "{\\"message\\":\\"hello world\\"}",
    },
    "cookies": Array [],
    "headers": Array [
      Object {
        "name": "x-powered-by",
        "value": "Express",
      },
      Object {
        "name": "content-type",
        "value": "application/json; charset=utf-8",
      },
      Object {
        "name": "content-length",
        "value": "25",
      },
      Object {
        "name": "etag",
        "value": "W/\\"19-sk8ULQY45ndZU9lmLRGpc9jYWFo\\"",
      },
    ],
    "headersSize": 0,
    "httpVersion": "",
    "redirectURL": "",
    "status": 200,
    "statusText": "OK",
  },
}
`;
